### PR TITLE
Adding Adams Bashforth time integrator [adams-bashforth-dev]

### PR DESCRIPTION
**Pull request by IdoAkkerman**
This PR adds  Adams Bashforth time integration. 
I also added a integration selector function, this could help  uniformising integration selection across the examples and miniapps.

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/513 | @IdoAkkerman  | @tzanio | @cjvogl + @delyank | 11/23/19 | 01/16/20 | ⌛due 01/23/20 |

### CODE REVIEW

**Comment by IdoAkkerman** [comment]
I also added Adams Moulton integration

**Comment by tzanio** [comment]
Please also update `CHANGELOG`

**Comment by tzanio** [comment]
Reminder: the reviews for this PR are due this week

cc: @cjvogl, @delyank

**Comment by delyank** [comment]
I'm overall positive about this pull request. It is a low-risk code change, since it only adds new functionality and does not modify existing one. Also, having these methods can be useful for the MFEM users. Code seems clean.

I also would like to have a test example both to demonstrate the method to potential users and to verify its accuracy now and in the future. Either a convergence test or a comparison test with another method would be good.

**Comment by tzanio** [comment]
Pushed back to after the holidays while waiting for input from @IdoAkkerman...

**Comment by tzanio** [comment]
Happy new year @IdoAkkerman! Please let us know when you will be able to take a look at this...

**Comment by IdoAkkerman** [comment]
Wanted to work on over the holidays, but did not find time. Will work on it now...

**Comment by IdoAkkerman** [comment]
Happy new year!

**Comment by IdoAkkerman** [comment]
I added a unit test, and it turns out AdamsBashforth, AdamsMoulton and Generalized alpha timeintegrators are only first order.

I have no clue so far as to why...

**Comment by IdoAkkerman** [comment]
Good news is that the review process is doing its job!

**Comment by tzanio** [comment]
Thank you @IdoAkkerman !

**Comment by tzanio** [comment]
@cjvogl, can you help @IdoAkkerman figure out why he is getting these convergence rates?

**Comment by cjvogl** [comment]
@tzanio, I’m on travel and about to lose email access, but will dive into this when I’m back next week.

**Comment by tzanio** [comment]
Adding additional week to figure out convergence issue

**Comment by kmittal2** [comment]
> I added a unit test, and it turns out AdamsBashforth, AdamsMoulton and Generalized alpha timeintegrators are only first order.
> 
> I have no clue so far as to why...

Have you made sure that the solution is bootstrapped properly at the beginning of the time-stepping process? For example, for AB2 you need to either provide solution at the first two steps to initialize the run OR take the first step with AB1 with a time-step size that is of length dt/2.

**Comment by IdoAkkerman** [comment]
Thanks for the comment. I figured it out independently, but it must indeed bethe start up. 
I fixed it for gen-alpha, will work on AB and AM tomorrow.

Thank you.

**Comment by IdoAkkerman** [comment]
I added a unit test for all time integrators.
Adding an example would be trivial? Is this still desired?

**Comment by cjvogl** [comment]
> Adding an example would be trivial? Is this still desired?

I'm ok without the example but @delyank might still want one.

**Comment by IdoAkkerman** [comment]
Personally, I am also wondering if the combination explicit/implicit solves is problematic.
For instance also in the bootstrap of the gen-alpha time integrator. For instance with Navier-Stokes a explicit solver might not be available. 
I am not to familiar concerning A and L stability, so I don't mind making changes (and learning!! thanks!) but I would need some pointers.

**Comment by IdoAkkerman** [comment]
Regarding the underscoring comment. I copied the syntax from the RK methods.
I personally don't like the underscores, but do see the need, therefore I prefer to use them for the function argument in the constructor. My guess was that that was the MFEM convention.

When looking at the ode.hpp file only the SIA class has underscores for the member function. Moreover, they list the functions before the members, which also seems at odds with the MFEM practices.

**Comment by IdoAkkerman** [comment]
Regarding stability in combination with the start-up.
AB can not be A-stable.
AM can be A-stable for  p<=2. that is AM0 and AM1.
In AM0 and AM1 no bootstrapping is required.
I made 2 changes:
1) For AM the IC is  converted to a history parameter. This reduces the bootstrapping by 1 step. Hence no bootstrapping for AM1.
2) The other AM methods are bootstrapped with SDIRK methods.

I think this solves the bootstrapping issues.
Explicit AB is bootstrapped with Explicit RK of equal convergence order
Implicit AM is bootstrapped with Implicit SDIRK of equal convergence order
Additionally, AM/AB are based on "extrapolotation" and RK/SDIRK on "interpolation" so I **assume** that  stability  and absolute accuracy are at least as good as the AM/AB methods themselves.

**Comment by delyank** [comment]
> I'm ok without the example but @delyank might still want one.

I'm happy as long as there is code that uses the new feature. I'm also satisfied with the solution of the bootstrapping issues. It seems that some of the tests fail. This needs to be addressed.

Maybe this is not a question for this particular pull request, but I'm going to ask it. To what extent do we want the bootstrapping to be integrated (hard-coded) within the overall time-stepping algorithm vs allowing more freedom of choice to the user in the interface, while providing reasonable default behaviors (choices)?

**Comment by cjvogl** [comment]
> Regarding stability in combination with the start-up.
> AB can not be A-stable.
> AM can be A-stable for p<=2. that is AM0 and AM1.
> In AM0 and AM1 no bootstrapping is required.
> I made 2 changes:
> 
>     1. For AM the IC is  converted to a history parameter. This reduces the bootstrapping by 1 step. Hence no bootstrapping for AM1.
> 
>     2. The other AM methods are bootstrapped with SDIRK methods.
> 
> 
> I think this solves the bootstrapping issues.
> Explicit AB is bootstrapped with Explicit RK of equal convergence order
> Implicit AM is bootstrapped with Implicit SDIRK of equal convergence order
> Additionally, AM/AB are based on "extrapolotation" and RK/SDIRK on "interpolation" so I **assume** that stability and absolute accuracy are at least as good as the AM/AB methods themselves.

Good point on AM0 and AM1 not requiring bootstrapping.  I'm no expert on linear multistep methods, but from the stability regions I looked up, your assumption holds: the n-th order AB stability region is a subset of the n-th order RK region and the n-th order AM region is a subset of the n-th order SDIRK region.  Given that, and the change to using SDIRK methods for the AM bootstrapping, I think my concern over the bootstrapping is alleviated.

That all being said, I do think @delyank's suggestion of having the user compute the initial solution staqes with whatever means is best for their application (in terms of stability, accuracy, cost, etc) and providing that as input to the solver is one that is worth consideration.

**Comment by cjvogl** [comment]
> Regarding the underscoring comment. I copied the syntax from the RK methods.
> I personally don't like the underscores, but do see the need, therefore I prefer to use them for the function argument in the constructor. My guess was that that was the MFEM convention.
> 
> When looking at the ode.hpp file only the SIA class has underscores for the member function. Moreover, they list the functions before the members, which also seems at odds with the MFEM practices.

Ah, if there is inconsistent underscore usage already present elsewhere in MFEM, then I guess there is no standard to adhere to here.  For what it's worth, I've found using
``` c++
void myfunc(double a)
{ this->a = a; }
```
a nice way to avoid underscores (which I also personally dislike).

**Comment by tzanio** [comment]
Merged in `next` for testing ...

**Comment by tzanio** [comment]
The [internal regression testing](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/cd2df000c6ea794a5843f0a513804150695d2323#tux426/2020-01-17-llnl-gh-next/build-test-basic.err) found this error:

```
linalg/ode.cpp: In member function ‘virtual void mfem::AdamsBashforthSolver::Step(mfem::Vector&, double&, double&)’:
linalg/ode.cpp:381:27: warning: operation on ‘((mfem::AdamsBashforthSolver*)this)->mfem::AdamsBashforthSolver::s’ may be undefined [-Wsequence-point]
    s = std::min(++s, smax);
                           ^
linalg/ode.cpp: In member function ‘virtual void mfem::AdamsMoultonSolver::Step(mfem::Vector&, double&, double&)’:
linalg/ode.cpp:449:27: warning: operation on ‘((mfem::AdamsMoultonSolver*)this)->mfem::AdamsMoultonSolver::s’ may be undefined [-Wsequence-point]
    s = std::min(++s, smax);
                           ^
```

Can you please address?

File: tests/unit/linalg/test_ode.cpp [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+//
+
+#include "mfem.hpp"
```
Comment (tzanio) [inline tests/unit/linalg/test_ode.cpp]: Thanks for including this @IdoAkkerman!

**Comment by tzanio** [comment]
> The [internal regression testing](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/cd2df000c6ea794a5843f0a513804150695d2323#tux426/2020-01-17-llnl-gh-next/build-test-basic.err) found this error:
> 
> ```
> linalg/ode.cpp: In member function ‘virtual void mfem::AdamsBashforthSolver::Step(mfem::Vector&, double&, double&)’:
> linalg/ode.cpp:381:27: warning: operation on ‘((mfem::AdamsBashforthSolver*)this)->mfem::AdamsBashforthSolver::s’ may be undefined [-Wsequence-point]
>     s = std::min(++s, smax);
>                            ^
> linalg/ode.cpp: In member function ‘virtual void mfem::AdamsMoultonSolver::Step(mfem::Vector&, double&, double&)’:
> linalg/ode.cpp:449:27: warning: operation on ‘((mfem::AdamsMoultonSolver*)this)->mfem::AdamsMoultonSolver::s’ may be undefined [-Wsequence-point]
>     s = std::min(++s, smax);
>                            ^
> ```
> 
> Can you please address?

Let me know if https://github.com/mfem/mfem/pull/513/commits/4e1a5465b09ffd47af92635b6cf2d91b169aa645 is the right fix

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by tzanio** [comment]
Turned out that problem was that `s` was modified twice, see https://github.com/mfem/mfem/pull/513/commits/9794864901687cb83ac39657dcdb85914a389817

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Pull request merged**